libraries/chain/chain_controller.cpp

```diff
@@ -344,32 +306,18 @@ transaction_trace chain_controller::_push_transaction(const packed_transaction&
 
 } FC_CAPTURE_AND_RETHROW( (transaction_header(packed_trx.get_transaction())) ) }
 
-static void record_locks_for_data_access(transaction_trace& trace, flat_set<shard_lock>& read_locks, flat_set<shard_lock>& write_locks ) {
-   for (const auto& at: trace.action_traces) {
-      for (const auto& access: at.data_access) {
-         if (access.type == data_access_info::read) {
-            trace.read_locks.emplace(shard_lock{access.code, access.scope});
-         } else {
-            trace.write_locks.emplace(shard_lock{access.code, access.scope});
-         }
-      }
-   }
-
-   // remove read locks for write locks taken by other actions
-   std::for_each(trace.write_locks.begin(), trace.write_locks.end(), [&]( const shard_lock& l){
-      trace.read_locks.erase(l); read_locks.erase(l);
-   });
-
-   read_locks.insert(trace.read_locks.begin(), trace.read_locks.end());
-   write_locks.insert(trace.write_locks.begin(), trace.write_locks.end());
-}
-
 transaction_trace chain_controller::_push_transaction( transaction_metadata&& data )
 { try {
    auto process_apply_transaction = [this](transaction_metadata& meta) {
       auto cyclenum = _pending_block->regions.back().cycles_summary.size() - 1;
       //wdump((transaction_header(meta.trx())));
 
+      const auto& trx = meta.trx();
+
+      // Validate uniqueness and expiration again
+      validate_uniqueness(trx);
```

Comment:
is this only here to cover the execution path where we extract a `generated_transaction_object` into a `transaction_metadata` and call `_push_transaction` ?

If so, can we create a legit `_push_transaction(const generated_transaction_object& )` and put this logic in there as well as any other stuff like `transaction_metadata` creation?

Reply:
Yes, that is exactly what it is for.

The way it is coded up now it does mean that a redundant check for uniqueness and  expiration will happen for undelayed input transactions. The `_push_transaction(const generated_transaction_object& )` overload you suggest is a way to get around that, but I think it is okay for the time being and we should delay that change until we eventually refactor the current code in chain_controller.cpp to be much easier to follow.